### DISCUSSION TITLE
Hard fork planning

### DISCUSSION THREAD

**Discussion by smichel17** [PaperWM-community/PaperWM#27] [General]
I'm imagining hard forking will look something like this:

- Merge https://github.com/PaperWM-community/PaperWM/pull/25 and the other PRs it links.
- Update the explanation in the README.
- Update the installation instructions and other links to the repo to point here.
- Change the extension namespace. (To what?)
    - Distribute on extensions.gnome.org (if someone wants to do this). 
- Switch to a standard GitHub / trunk-based branching model:
    - `master` is the latest code that everything forks from. All PRs go here.
    - Releases are git tags.
- Enable issues for this repo.
    - If anyone wants to migrate issues from the old repo, they're welcome (as long as the issues can be reproduced here). Anyone who does this: please link back to the old issue, that way we have a way to keep track of which issues have been migrated or not.
- Support only the latest gnome-shell version, to start.
    - If it turns out we have enough developer time and interest to support older versions, we can do that later (and consider how it might affect our git branch flow).
    - Doesn't *necessarily* mean removing all the existing compatibility code… but that's something for the people actually doing the coding to decide (whether it's getting in the way), not me.
- Go through the other PRs and try to get them either merged or closed.

Before we start, I think we should try to agree on some minimal governance. Not too much or _too detailed_, just who gets a say and how it gets counted, about these decisions:

- **Security:** Grant someone access. Revoke that access. Grant ownership. Revoke ownership.
- **Features:** Reject a feature request. Reject a PR. Approve a PR.
    - "Who should merge the PR and when?" is a good example of *too detailed*. If someone merges a broken PR, that's easy to fix, as long as we already decided to merge (eventually). Finding out afterward that some relevant person was opposed and was left out of the discussion… that's what we're trying to avoid; broken trust is much harder to rebuild.

https://communityrule.info/ is a good reference with vocabulary for thinking about those questions clearly. Depending on the answers, we'll likely modify the plan above.

**Comment by michaelhthomas** [comment]
+1 for switching to a standard trunk-based branching model, and hopefully adding some policies to standardize some of the branching and PR approval processes. We can probably create a new `main` branch from the existing `next-release`, and then merge in all the GNOME 42 PRs there.

**Reply by michaelhthomas** [reply]
Here's sort of what I'm thinking from a branch protection standpoint, to govern how pull requests get merged (and require multiple reviews) so that there's a standardized approval process
![Screenshot 2022-06-24 at 15-56-13 Manage access · PaperWM-community_PaperWM](https://user-images.githubusercontent.com/18223295/175658892-da208256-0621-4481-86e6-89cafd891cd4.png)

**Reply by smichel17** [reply]
- I think 1 approval required is probably fine to start. If bad code is getting merged, we can re-evaluate and bump it up later.

- I'm not sure about `Require linear history`. I generally prefer merge commits > rebase > squash.
    - Squash loses information about what order things happened in (if you are committing early and often).
    - Having a commit lets you tie it to the PR, which sometimes contains important context (even though that *should* be in the commit messages). It can make the history a bit harder to navigate if you're not fluent in git, though.
        - Note: you can manually rebase the branch before merging the PR using a merge commit, to get both a ~linear history _and_ merge commit messages referencing the PR.

Everything else looks good to me. And it can always be tweaked as we go.

**Reply by michaelhthomas** [reply]
Agreed on both fronts (I didn't even realize require linear history was selected, oops), and there definitely will be some tweaking along the way.

**Reply by smichel17** [reply]
I set it up like above, including the two changes I suggested, and one other:

- Enabled `Restrict who can push to matching branches`

   It was a little difficult for me to find out exactly what this *does*, but I discovered that it is only available for organizations, not repos of an individual. So, I it prevents people who are members but not admins of the org from merging. Or, as the setting puts it, allow merging only by "organization administrators, repository administrators, and users with the Maintain role". I believe those groups still need to use PRs (i.e. it does not allow them to push directly), but I'm not 100% sure.

**Comment by smichel17** [comment]
My opinions:

## Security

> Grant someone access. Revoke that access. Grant ownership. Revoke ownership.

Okay, I guess there's really three levels of access to a repo that matter:

<details><summary><b>Maintainer</b> role on the <i>repository</i>. I suggest granting this to <b>any contributor</b> who gets a non-trivial PR merged, for ease of collaboration.</summary>

- What it grants: Can merge code to `master` through the normal PR process, but cannot bypass branch restrictions or give other people permissions on the repo. Can change a small number of "general settings" on the repo (e.g. which merge strategies are permitted).
    - This is all people need for day-to-day maintenance.
- Risks: There is really very little harm that someone with this role can do, intentionally or accidentally. Anything pushed to master can be reverted. I guess they could deface the issues, but that seems really unlikely; I've never seen someone inside a community do it.
</details>

<details><summary><b>Admin</b> role on the <i>repository</i>. I think this should be the "<b>core dev team</b>". Something like: New people are added by majority vote of the existing team, and are removed either by majority vote or if they don't interact with the repo for a year or longer.</summary>

- What: Can change all repo settings and add other collaborators. There is an org-level setting to make it so admins can't delete repos, though.
- Risks: Could accidentally mess up the git history, or sneak commits into `master` without going through the PR process (maliciously, or more likely accidentally pushing a broken commit). Overall still pretty low risk, since the nature of git makes it likely someone else will have a copy.
</details>


<details><summary><b>Owner</b> role on the <i>organization</i>. I suggest that <b>2-3 people</b> (1 active developer and 1-2 backups) should have this permission, basically <i>just to ensure we don't lose access to the org</i> again, like we did with @paperwm.  I'm happy to be a backup.</summary>

- What: Can delete the repo and change org-wide settings.
    - Most of the time, a repo admin could do anything needed; having this role doesn't enable anything more.
- Risks: Could kick everyone else out, claiming control over the namespace. Could [accidentally delete the repo](https://httpie.io/blog/stardust).

I would view this role as an implementation detail, not part of the governance; whoever has it wouldn't get any special voting privileges. So the only requirement for this position is to be trustworthy enough not to run off with the repo :)
</details>

## Features

> Reject a feature request.

When it comes to keeping a focused vision for the design and avoiding feature creep, I think it's hard for a group/committee to do that. Instead, I would suggest appointing one person to be the arbiter of "does this feature belong in PaperWM?" That person doesn't have to make all the decisions themselves, just set the standards.

Choosing this person is probably the hardest decision. Most FLOSS projects default to "whoever started the project" (or the most active current maintainer) in a BDFL type model. I *can* do it if nobody else wants to, but honestly I have enough other projects right now, I'd prefer to step away from PaperWM development once the fork is running smoothly.

> Reject a PR.

Same for coding standards, although it doesn't need to be the same person. I think this should be the most active maintainer.

> Approve a PR.

On the other hand, I think it's good to be permissive here, and say that if a PR is in scope and good code according to the standards above, anyone with the `Admin` or `Maintainer` role should be able to approve/merge it. If they judged it wrong, it's easy enough to revert.

**Reply by smichel17** [reply]
Sorry this got so long. I could probably make it 1/3 the length if I wanted to edit it instead of brain-dump, but I'm erring on the side of spending less time writing and sending it sooner.

edit: Reformatted it to hide less important details.

**Comment by michaelhthomas** [comment]
@smichel17 What are your thoughts about merging #25 into `prettier`, and then just renaming the `prettier` branch to `main` and marking it as the default? (This is assuming that the `prettier` branch is currently up to date with `next-release`, but I believe that's the case.

**Reply by smichel17** [reply]
Does GitHub automatically point PRs at the new branch if you rename it? If so, it might be worth renaming `next-release` instead, so the PRs don't have to be closed and re-opened pointing at the new branch. Otherwise, I don't think it matters that much which branch is renamed.

**Reply by michaelhthomas** [reply]
It'll sort of be a challenge to deal with, since the gnome 42 PR as it stands currently depends on the repo being prettier formatted, and is thus targeted against the prettier branch. Prettier formatting will probably create merge conflicts with all the other open PRs, while not prettier formatting will make it a lot more challenging to merge the gnome 42 changes.

**Reply by michaelhthomas** [reply]
I think you're right, and it may be best to merge all the open PRs (that need merging), into `next-release`, rename it, rerun the prettier format, and then merge the GNOME 42 PR

**Reply by smichel17** [reply]
@ccope Your #25 includes a bunch of different PRs. Do you want to close those PRs and just merge #25, or merge those PRs and then rebase #25?

**Reply by smichel17** [reply]
In the interest of getting things done while I have time to spend, I'm going to go ahead and merge the PRs. If this turns out to create problems, `next-release` is currently at https://github.com/PaperWM-community/PaperWM/commit/2852073861f7dcc4e8c660286ebf3c597443afc4.

**Reply by smichel17** [reply]
> Does GitHub automatically point PRs at the new branch if you rename it?

I tested in a different repo, and it does. So I'm going to rename `next-release` to `master` now, make it the default branch, and remove the old `next-release` and `develop` branches, which will now be fully merged to `master`.

**Reply by smichel17** [reply]
I'm done merging PRs for now / today. Unfortunately the rename also closed the PR back to the main repo, so I also restored the `next-release` branch (as a new branch) as it was, to keep that PR open, for visibility. At some point in the future we can delete the `next-release` branch but for now I guess we can just leave it collecting dust. It's no longer the default branch anyway.

**Comment by smichel17** [comment]
Maybe an example will be more useful than my [other comment](https://github.com/PaperWM-community/PaperWM/discussions/27#discussioncomment-3044254).

Let's pretend I'm against @ccope's PR to format the repo with Prettier (https://github.com/PaperWM-community/PaperWM/pull/26). How to decide whether to use Prettier or not?

**Reply by michaelhthomas** [reply]
I personally like the idea of having a single person be appointed as a sort of "project manager" that gets to decide which features are necessary. I think that role would include deciding whether or not to preserve PRs like the "prettier format the repository" PR that make sweeping changes.

**Reply by smichel17** [reply]
Who appoints them? Single person, group of developers (anyone involved, not just programmers), or community at large (users)?

**Reply by michaelhthomas** [reply]
Given the size of the community, I think if someone stands up and there are no strong objections, then that would be enough. And if that person were to become unavailable or otherwise unwanted, then the community of developers and users on this repo could decide on a suitable replacement.

**Reply by smichel17** [reply]
That seems reasonable, and pretty standard. Essentially, "prioritize getting the work done and not fragmenting the community."

I'd add, this role shouldn't be able to force anyone to implement anything, only decide what should/shouldn't be accepted. That's *kind of* stating the obvious in a volunteer project, but I think it's good to be explicit.

I suppose I'll start in that role since nobody else has volunteered yet. Long-term I'd prefer to have someone else replace me.

**Reply by michaelhthomas** [reply]
Thanks for all your work!

**Comment by smichel17** [comment]
I've gone ahead and enabled issues in this repo

**Comment by smichel17** [comment]
I guess there is one big question:

- **https://github.com/PaperWM-community/PaperWM/discussions/30**

On a related subject (platforms) I suggest that we should **not** set up a chat room anywhere. The community isn't big enough to need it yet, and having all communication in a place that's relatively easy to search and link back to is really convenient for any kind of project without people working full-time who you can just ask if you need context on something.

**Comment by smichel17** [comment]
Update: someone finally managed to get in touch with one of the PaperWM maintainers, and I was granted maintainer access to the main PaperWM repo today. So, I think we'd do best to plan on moving back to that repo. I'll post more thoughts over in https://github.com/paperwm/PaperWM/issues/407 when I get the chance— probably tomorrow or this weekend, but I'm just a little drained at the moment.

tl;dr I only have maintainer access there so I can't add other people, so I propose we basically follow the process above but I'll manually merge PRs after someone else has approved them.

**Reply by michaelhthomas** [reply]
Sweet, thanks for all your work!
